README.md
# BusDroid RJ
### Um simples aplicativo que permite localizar as linhas de ônibus disponíveis no município do Rio de Janeiro!

### Descrição:

###### Este é um simples aplicativo escrito em Java e feito para a plataforma Android API 14 (Android 4.0 em diante) que possui a função mostrar a localização das linhas de ônibus disponíveis no município do Rio de Janeiro, utilizando a API do data.rio para recuperar as informações dos ônibus em tempo real e utilizando a API do [Google Maps Android API](https://developers.google.com/maps/documentation/android-api/?hl=pt-br) para indicar o local exato dos ônibus em um mapa interativo.

###### A tela principal do aplicativo mostra dois possíveis modos de busca dos ônibus disponíveis no momento: O modo de busca pela linha do ônibus e pelo número de ordem do ônibus. 

###### Ao buscar pela linha, o aplicativo irá indicar todos os ônibus daquela linha que estão circulando. Ao buscar pelo número de ordem, o aplicativo irá indicar apenas o ônibus que possuir aquele número.

###### O aplicativo também poderá mostrar a localização atual do usuário no mapa, facilitando a identificação visual dos ônibus que estejam mais próximos ao usuário. Para habilitar esta função, é necessário conceder permissões de localização ao aplicativo e ligar a opção de Localização nas configurações do Android.
###### A função de localização é opcional, não sendo necessária para que o aplicativo funcione de maneira correta!

###### Ao buscar por uma linha de ônibus ou um ônibus específico através de seu número de ordem, uma janela com o Google Maps irá abrir e indicar através de marcadores as localizações dos veículos disponíveis no momento.

###### Os marcadores irão conter informações como a linha do ônibus, o número de ordem, a hora que o GPS do ônibus foi atualizado com o sistema e a velocidade do veículo naquele momento.

##### Abaixo está uma demonstração de busca por uma linha de ônibus específica:

<img src="http://i.imgur.com/TvKtSGF.png" width="100%" height="100%" />
<img src="http://i.imgur.com/QnkzcVG.png" width="100%" height="100%" />

## [Verifique o CHANGELOG.txt para maiores informações sobre novas versões!](https://raw.github.com/Wolfterro/BusDroid-RJ/master/CHANGELOG.txt)

### Idiomas Disponíveis:
 - Português Brasileiro (Padrão)
 - Inglês
 
 ### Permissões Necessárias:
 - Localização (Opcional)

### Download:

##### Para baixar o código-fonte do aplicativo, basta utilizar o git para clonar o repositório:
    git clone https://github.com/Wolfterro/BusDroid-RJ.git
    cd BusDroid-RJ/

##### Caso queira baixar o aplicativo já compilado, em formato .apk, use o link abaixo:
#### ***Release:*** https://github.com/Wolfterro/BusDroid-RJ/releases/tag/v1.1-Release

#### Atenção!
##### Caso queira compilar o aplicativo por conta própria, terá que obter uma chave da API do [Google Maps Android API](https://developers.google.com/maps/documentation/android-api/?hl=pt-br) vinculada a sua chave de assinatura do aplicativo! 
##### Caso contrário, o mapa do aplicativo NÃO IRÁ FUNCIONAR!

##### Caso não possua o git e queira baixar o repositório, baixe através deste [Link](https://github.com/Wolfterro/BusDroid-RJ/archive/master.zip) ou clique em "Clone or Download", no topo da página.


CHANGELOG.txt
BusDroid RJ
===========

Um simples aplicativo que permite localizar as linhas de ônibus no município do Rio de Janeiro!
-----------------------------------------------------------------------------------------------

CHANGELOG:
==========

VERSÃO 1.1:
- Alterado layout da janela de informações do marcador, permitindo adicionar mais informações futuramente.
- Adicionado trajeto das linhas de ônibus na janela de informações do marcador.
- Adicionada novas informações sobre o aplicativo no menu "Sobre".
- Pequenas alterações de layout e bugfixes.

#-------------------------------------------------------------------------------------------------------------------

VERSÃO 1.0:
- Versão inicial do aplicativo.

#-------------------------------------------------------------------------------------------------------------------

app/src/main/java/io/github/wolfterro/busdroidrj/GetCSV.java
/*
The MIT License (MIT)

Copyright (c) 2017 Wolfgang Almeida

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

package io.github.wolfterro.busdroidrj;

import android.content.Context;

import java.net.*;
import java.io.*;

import com.opencsv.CSVReader;

public class GetCSV extends Thread {

    // Propriedades privadas
    // =====================
    private Context c = null;

    private String busLine = "";
    private String route = "";

    // Construtor da classe
    // ====================
    public GetCSV(Context c, String busLine) {
        this.c = c;

        this.busLine = busLine;
    }

    // Métodos privados da classe
    // ==========================

    // Iniciando resgate de valores sobre a linha de ônibus escolhida
    // ==============================================================
    @Override
    public void run() {
        try {
            // Retirando as letras da linha atual do ônibus
            // --------------------------------------------
            String busLineNoLetters = busLine.replaceAll("[^\\d]", "");

            URL u = new URL(String.format(GlobalVars.busRouteAPI, busLineNoLetters));
            BufferedReader in = new BufferedReader(new InputStreamReader(u.openStream()));

            CSVReader reader = new CSVReader(in);

            // Adicionando zeros para a busca de linhas com menos de 3 números
            // ---------------------------------------------------------------
            if(busLineNoLetters.length() == 1) {
                busLineNoLetters = String.format("00%s", busLineNoLetters);
            }
            else if(busLineNoLetters.length() == 2) {
                busLineNoLetters = String.format("0%s", busLineNoLetters);
            }

            String[] nextLine;
            while ((nextLine = reader.readNext()) != null) {

                // Imprime o trajeto da linha de ônibus, sem a linha atual do ônibus ao lado
                // -------------------------------------------------------------------------
                if(nextLine[1].startsWith(busLineNoLetters)) {
                    route = nextLine[1].replace(busLineNoLetters + "-", "");
                    break;
                }
                else {
                    route = c.getString(R.string.notAvailable);
                    continue;
                }
            }

            reader.close();
        }
        catch(MalformedURLException e) {
            e.printStackTrace();
            route = c.getString(R.string.notAvailable);
        }
        catch(IOException e) {
            e.printStackTrace();
            route = c.getString(R.string.notAvailable);
        }
    }

    // Métodos públicos da classe
    // ==========================

    // Resgatando o trajeto da linha de ônibus
    // =======================================
    public String getRoute() {
        return route;
    }
}

app/src/main/java/io/github/wolfterro/busdroidrj/GetJSON.java
/*
The MIT License (MIT)

Copyright (c) 2017 Wolfgang Almeida

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

package io.github.wolfterro.busdroidrj;

import java.io.IOException;
import java.io.InputStream;
import java.net.*;
import java.util.Scanner;

import org.json.*;

// VERSÃO 1.0

// Classe responsável por resgatar o arquivo .json de uma API pública
// ==================================================================
public class GetJSON {
    // Propriedades privadas
    // =====================
    private JSONObject obj = null;
    private String apiURL = "";

    // Propriedades públicas
    // =====================
    public String errorMsg = "";

    // Construtor da classe
    // ====================
    public GetJSON(String apiURL) {
        this.apiURL = apiURL;
    }

    // Métodos privados da classe
    // ==========================
    private void getJSON() {
        try {
            URL u = new URL(apiURL);
            HttpURLConnection conn = (HttpURLConnection) u.openConnection();
            conn.connect();

            InputStream stream = (InputStream) conn.getContent();

            Scanner s = new Scanner(stream);
            s.useDelimiter("\\A");
            String res = s.hasNext() ? s.next() : "";
            s.close();

            obj = new JSONObject(res);
            errorMsg = "OK!";
        }
        catch(MalformedURLException e) {
            e.printStackTrace();
            errorMsg = "MALFORMED_URL_EXCEPTION!";
        }
        catch(IOException e) {
            e.printStackTrace();
            errorMsg = "NOT_FOUND_OR_INACCESSIBLE!";
        }
        catch(org.json.JSONException e) {
            e.printStackTrace();
            errorMsg = "JSON_EXCEPTION_ERROR_GETJSON!";
        }
        catch(Exception e) {
            e.printStackTrace();
            errorMsg = "GENERIC_EXCEPTION_ERROR!";
        }
    }

    // Métodos públicos da classe
    // ==========================
    public JSONObject get() {
        getJSON();
        return obj;
    }

    public String getErrorMsg() {
        return errorMsg;
    }
}

app/src/main/java/io/github/wolfterro/busdroidrj/GlobalVars.java
/*
The MIT License (MIT)

Copyright (c) 2017 Wolfgang Almeida

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

package io.github.wolfterro.busdroidrj;

/**
 * Created by Wolfterro on 16/07/2017.
 */

public class GlobalVars {
    public static String busAPI = "http://dadosabertos.rio.rj.gov.br/apiTransporte/apresentacao/rest/index.cfm/obterTodasPosicoes.json";
    public static String busRouteAPI = "http://dadosabertos.rio.rj.gov.br/apiTransporte/Apresentacao/csv/gtfs/onibus/percursos/gtfs_linha%s-shapes.csv";
}


app/src/main/java/io/github/wolfterro/busdroidrj/MainActivity.java
/*
The MIT License (MIT)

Copyright (c) 2017 Wolfgang Almeida

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

package io.github.wolfterro.busdroidrj;

import android.Manifest;
import android.app.ProgressDialog;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    // Elementos da Activity Principal
    // ===============================
    private EditText editTextLine;
    private Button buttonLine;

    private EditText editTextOrder;
    private Button buttonOrder;

    // Propriedades Privadas da classe
    // ===============================
    private final int PERMISSION_GRANTED_VALUE = 0;

    // Menu de opções da Activity Principal
    // ====================================
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.option_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    // Selecionando opções no menu da Activity Principal
    // =================================================
    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        switch (item.getItemId()) {
            case R.id.about:
                Intent about = new Intent(MainActivity.this, AboutActivity.class);
                MainActivity.this.startActivity(about);

                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    // Criando Activity Principal do aplicativo
    // ========================================
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Seleção para a linha de ônibus
        // ------------------------------
        editTextLine = (EditText)findViewById(R.id.editTextBusLine);
        buttonLine = (Button)findViewById(R.id.buttonLine);

        // Seleção para a ordem do ônibus
        // ------------------------------
        editTextOrder = (EditText)findViewById(R.id.editTextBusOrder);
        buttonOrder = (Button)findViewById(R.id.buttonOrder);

        // Pedindo permissão do usuário para determinar sua localização
        // ------------------------------------------------------------
        if(ContextCompat.checkSelfPermission(MainActivity.this,
                Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {

            ActivityCompat.requestPermissions(MainActivity.this,
                    new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
                    PERMISSION_GRANTED_VALUE);
        }

        // Buscando ônibus através da linha inserida
        // -----------------------------------------
        buttonLine.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(editTextLine.getText().toString().equals("")) {
                    Toast.makeText(MainActivity.this,
                            getString(R.string.pleaseInsertBusLine),
                            Toast.LENGTH_SHORT).show();
                }
                else {
                    // Iniciando o resgate de informações
                    // ==================================
                    ProgressDialog pd = new ProgressDialog(MainActivity.this);
                    pd.setTitle(getString(R.string.gettingInformations));
                    pd.setMessage(getString(R.string.pleaseWait));
                    pd.setCancelable(false);
                    pd.show();

                    GetBusLineThread gb = new GetBusLineThread(editTextLine.getText().toString(),
                            MainActivity.this,
                            pd);

                    gb.start();
                }
            }
        });

        // Buscando ônibus através da ordem inserida
        // -----------------------------------------
        buttonOrder.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if(editTextOrder.getText().toString().equals("")) {
                    Toast.makeText(MainActivity.this,
                            getString(R.string.pleaseInsertBusOrder),
                            Toast.LENGTH_SHORT).show();
                }
                else {
                    // Iniciando o resgate de informações
                    // ==================================
                    ProgressDialog pd = new ProgressDialog(MainActivity.this);
                    pd.setTitle(getString(R.string.gettingInformations));
                    pd.setMessage(getString(R.string.pleaseWait));
                    pd.setCancelable(false);
                    pd.show();

                    GetBusOrderThread gb = new GetBusOrderThread(editTextOrder.getText().toString(),
                            MainActivity.this,
                            pd);

                    gb.start();
                }
            }
        });
    }
}


app/src/main/java/io/github/wolfterro/busdroidrj/MapsActivity.java
/*
The MIT License (MIT)

Copyright (c) 2017 Wolfgang Almeida

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

package io.github.wolfterro.busdroidrj;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.support.v4.app.FragmentActivity;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.view.View;
import android.widget.TextView;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;

import java.util.ArrayList;
import java.util.Random;

public class MapsActivity extends FragmentActivity implements OnMapReadyCallback {

    // Propriedades da Activity
    // ========================
    private GoogleMap mMap;

    private String busLine = "";
    private ArrayList<String> results = new ArrayList<String>();
    private LatLng busOrderLocation = null;
    private String route = "";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_maps);

        Intent i = getIntent();
        busLine = i.getStringExtra("BUSLINE");
        results = i.getStringArrayListExtra("RESULTS");
        route = i.getStringExtra("ROUTE");

        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager()
                .findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        mMap = googleMap;

        // Determinando a localização do usuário se houver permissão
        // ---------------------------------------------------------
        if(ContextCompat.checkSelfPermission(MapsActivity.this,
                Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
            mMap.setMyLocationEnabled(true);
        }

        // Adicionando marcadores customizados no mapa
        // -------------------------------------------
        for(int x = 0; x < results.size(); x++) {
            Random random = new Random();   // So much random...

            String[] info = results.get(x).split(" -//- ");
            String ll = info[0];
            String order = info[1];
            String lastUpdate = info[2];
            String speed = info[3];

            double lat = Double.parseDouble(ll.split(",")[0]);
            double lng = Double.parseDouble(ll.split(",")[1]);

            LatLng loc = new LatLng(lat, lng);

            int selectedColor = random.nextInt(360);

            MarkerOptions options = new MarkerOptions();
            options.position(loc);
            options.title(String.format("%s - %s", busLine, order));
            options.snippet(String.format("%s: %s\n%s: %s\n%s: %s %s",
                    getString(R.string.route),
                    route,
                    getString(R.string.lastUpdate),
                    lastUpdate,
                    getString(R.string.speed),
                    speed,
                    getString(R.string.kmh)));
            options.icon(BitmapDescriptorFactory.defaultMarker(selectedColor));

            Marker marker = mMap.addMarker(options);

            mMap.setInfoWindowAdapter(new GoogleMap.InfoWindowAdapter() {
                @Override
                public View getInfoWindow(Marker marker) {
                    return null;
                }

                @Override
                public View getInfoContents(Marker marker) {
                    View contentView = getLayoutInflater()
                            .inflate(R.layout.custom_markerinfo, null);

                    TextView tvTitle = (TextView)contentView.findViewById(R.id.titleInfoWindow);
                    TextView tvSnippet = (TextView)contentView.findViewById(R.id.snippetInfoWindow);

                    tvTitle.setText(marker.getTitle());
                    tvSnippet.setText(marker.getSnippet());

                    return contentView;
                }
            });

            busOrderLocation = loc;
        }

        if(results.size() == 1) {
            mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(busOrderLocation, 16.0f));
        }
        else {
            LatLng rio = new LatLng(-22.95, -43.40);
            mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(rio, 10.0f));
        }
    }
}


app/src/main/java/io/github/wolfterro/busdroidrj/AboutActivity.java
/*
The MIT License (MIT)

Copyright (c) 2017 Wolfgang Almeida

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

package io.github.wolfterro.busdroidrj;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;

public class AboutActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_about);
    }
}


app/src/main/java/io/github/wolfterro/busdroidrj/GetBusLineThread.java
/*
The MIT License (MIT)

Copyright (c) 2017 Wolfgang Almeida

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

package io.github.wolfterro.busdroidrj;

/**
 * Created by Wolfterro on 16/07/2017.
 */

import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.os.Message;
import android.widget.Toast;

import org.json.*;

import java.util.ArrayList;

public class GetBusLineThread extends Thread {
    // Propriedades privadas
    // =====================
    private Context c = null;
    private ProgressDialog pd = null;

    private String busLineReceived = "";
    private Object busLine = null;

    private static int DATAHORA = 0;
    private static int ORDEM = 1;
    private static int LINHA = 2;
    private static int LATITUDE = 3;
    private static int LONGITUDE = 4;
    private static int VELOCIDADE = 5;

    private ArrayList<String> results = new ArrayList<String>();
    private boolean isOK = false;
    private boolean getInfoStatus = false;
    private String toastErrorMsg = "";

    // Construtor da classe
    // ====================
    public GetBusLineThread(String busLineReceived, Context c, ProgressDialog pd) {
        this.busLineReceived = busLineReceived;
        this.c = c;
        this.pd = pd;
    }

    // Métodos públicos
    // ================

    @Override
    public void run() {
        try {
            busLine = Double.parseDouble(busLineReceived);
        }
        catch(Exception e) {
            busLine = busLineReceived.toUpperCase();
        }

        getBusLineInfo();

        if(getInfoStatus) {
            if(results.size() == 0 || results == null) {
                isOK = false;
            }
            else {
                isOK = true;
            }
        }

        handler.sendEmptyMessage(0);
    }

    // Metodos privados
    // ================

    // Resgatando as informações da linha de ônibus escolhida
    // ======================================================
    private void getBusLineInfo() {
        GetJSON gj = new GetJSON(GlobalVars.busAPI);
        JSONObject jo = gj.get();

        if(!gj.getErrorMsg().equals("OK!")) {
            toastErrorMsg = String.format("%s\n%s %s",
                    c.getString(R.string.errorObtainingInformations),
                    c.getString(R.string.error),
                    gj.getErrorMsg());

            toastError.sendEmptyMessage(0);
            getInfoStatus = false;

            return;
        }

        try {
            JSONArray data = jo.getJSONArray("DATA");

            for(int i = 0; i < data.length(); i++) {
                JSONArray dataArray = data.getJSONArray(i);

                if(dataArray.get(LINHA).equals(busLine)) {
                    String latitLongit = String.format("%s,%s",
                            dataArray.get(LATITUDE),
                            dataArray.get(LONGITUDE));

                    String hora = dataArray.get(DATAHORA).toString().split(" ")[1];

                    String res = String.format("%s -//- %s -//- %s -//- %s",
                            latitLongit,
                            dataArray.get(ORDEM),
                            hora,
                            dataArray.get(VELOCIDADE));

                    results.add(res);
                }
            }

            getInfoStatus = true;
        }
        catch(org.json.JSONException e) {
            e.printStackTrace();

            toastErrorMsg = String.format("%s\n%s %s",
                    c.getString(R.string.errorObtainingInformations),
                    c.getString(R.string.error),
                    "JSON_EXCEPTION_ERROR_LINE!");

            toastError.sendEmptyMessage(0);
            getInfoStatus = false;
        }
    }

    // Dispensando a mensagem do ProgressDialog
    // ========================================
    private Handler handler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            if(getInfoStatus) {
                if(isOK) {
                    pd.dismiss();

                    String message = String.format(c.getString(R.string.foundXBusFromYLine),
                            results.size(),
                            busLineReceived);

                    Toast.makeText(c, message, Toast.LENGTH_LONG).show();

                    // Chamando o Activity do Google Maps
                    // ----------------------------------
                    Intent intent = new Intent(c, MapsActivity.class);
                    intent.putStringArrayListExtra("RESULTS", results);

                    String bLine = String.format("%s", busLine).replace(".0", "");
                    String markerTitle = String.format("%s %s",
                            c.getString(R.string.bus),
                            bLine);

                    intent.putExtra("BUSLINE", markerTitle);

                    GetCSV csv = new GetCSV(c, bLine);
                    csv.start();

                    try {
                        csv.join();
                        intent.putExtra("ROUTE", csv.getRoute());
                    }
                    catch (InterruptedException e) {
                        e.printStackTrace();
                        intent.putExtra("ROUTE", c.getString(R.string.notAvailable));
                    }

                    c.startActivity(intent);
                }
                else {
                    pd.dismiss();

                    String message = String.format(c.getString(R.string.noBusLineFound),
                            busLineReceived);
                    Toast.makeText(c, message, Toast.LENGTH_LONG).show();
                }
            }
            else {
                pd.dismiss();
            }
        }
    };

    // Mensagens de erro
    // =================
    private Handler toastError = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            Toast.makeText(c, toastErrorMsg, Toast.LENGTH_LONG).show();
        }
    };
}
